# Домашнее задание №2

**

## Часть 1. ML workflow (**всего 5 баллов**)

In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV

from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor

from sklearn.metrics import mean_squared_error, r2_score

from sklearn.preprocessing import PolynomialFeatures
from sklearn.pipeline import Pipeline

from itertools import product

### Загрузим данные для работы. 

In [2]:
df = pd.read_csv("winequality-red.csv")
df.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

Будем решать задачу регрессии: необходимо предсказать качество вина на основе его характеристик

### Шаг 1.  (**0.2 балла**)
Создайте матрицу X объект-признак и целевой вектор y ("quality")

In [3]:
y = df['quality'].to_numpy()
X = df.drop('quality', 1).to_numpy()

In [4]:
X

array([[ 7.4  ,  0.7  ,  0.   , ...,  3.51 ,  0.56 ,  9.4  ],
       [ 7.8  ,  0.88 ,  0.   , ...,  3.2  ,  0.68 ,  9.8  ],
       [ 7.8  ,  0.76 ,  0.04 , ...,  3.26 ,  0.65 ,  9.8  ],
       ...,
       [ 6.3  ,  0.51 ,  0.13 , ...,  3.42 ,  0.75 , 11.   ],
       [ 5.9  ,  0.645,  0.12 , ...,  3.57 ,  0.71 , 10.2  ],
       [ 6.   ,  0.31 ,  0.47 , ...,  3.39 ,  0.66 , 11.   ]])

In [5]:
y

array([5, 5, 5, ..., 6, 5, 6], dtype=int64)

### Шаг 2. (**0.2 балла**)
Разбейте данные на train и test (доля тестовых данных - 30%).

In [6]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=690)

Здесь просто чекаю, что датафрейм разделился корректно:

In [7]:
X_test.shape[0]/X.shape[0]

0.300187617260788

### Шаг 3. (**0.2 балла**)
Обучите линейную регрессию на тренировочных данных и сделайте предсказания на train и на test.

In [8]:
reg1 = LinearRegression().fit(X_train, y_train)
train_prediction = reg1.predict(X_train)
test_prediction = reg1.predict(X_test)

train_prediction

array([5.03329837, 5.87059011, 4.77301568, ..., 5.62725576, 6.17390055,
       6.34658728])

In [9]:
test_prediction

array([6.07333963, 5.16620447, 5.12337435, 6.71925388, 5.93382321,
       5.31450928, 6.29196632, 5.18222677, 5.50583533, 6.37922981,
       5.78198518, 5.13424372, 6.47034389, 5.1730555 , 6.31597327,
       5.42683583, 6.62169844, 5.26445765, 5.44373906, 5.22148189,
       6.0267704 , 5.47222617, 5.53749023, 5.6128583 , 5.26976299,
       5.35937958, 4.91041465, 5.37185731, 4.9377547 , 5.61046636,
       5.0946638 , 5.9715291 , 5.38884978, 5.08876448, 5.35035249,
       5.45307662, 5.30875147, 5.42296007, 5.69398686, 5.3284178 ,
       5.07124173, 5.02481411, 5.20276754, 6.3527944 , 5.69876373,
       5.43154818, 5.49229549, 5.12165396, 6.20242799, 5.11594365,
       5.23277114, 5.99794018, 5.17759795, 6.49994703, 5.82492147,
       4.84600398, 6.02752404, 6.41894492, 5.08806849, 6.21434959,
       5.62624554, 4.87483187, 5.87696915, 5.29044629, 4.94856693,
       5.4467288 , 5.86956752, 5.25035244, 6.93405126, 5.22739129,
       5.77171039, 5.525642  , 5.58295484, 5.56978658, 5.62870

### Шаг 4. (**0.4 балла**)
Выведите на экран ошибку MSE на train и на test, затем выведите на экран ошибку r2 на train и test.  

In [10]:
print('MSE на трейне =', mean_squared_error(y_train, train_prediction),'\nMSE на тесте = ', mean_squared_error(y_test, test_prediction))

MSE на трейне = 0.42157894227599524 
MSE на тесте =  0.4086296844296972


In [11]:
print('R-squared на трейне =', r2_score(y_train, train_prediction),'\nR-squared на тесте = ', r2_score(y_test, test_prediction))

R-squared на трейне = 0.3599876941197515 
R-squared на тесте =  0.35669534685378146


### Шаг 5. (**0.5 балла**)
Вычислите среднее качество (r2) модели на кросс-валидации с k=5 фолдами.

Качество на каждом фолде:

In [12]:
cross_val_score(reg1, X, y, cv=5, scoring='r2')

array([0.13200871, 0.31858135, 0.34955348, 0.369145  , 0.2809196 ])

Среднее качество по всем фолдам:

In [13]:
np.mean(cross_val_score(reg1, X, y, cv=5, scoring='r2'))

0.2900416288421999

### Шаг 6.  (**0.5 балла**)
Теперь примените линейную регрессию с L1-регуляризацией (Lasso) для данной задачи. Объявите модель и подберите параметр регуляризации alpha по сетке. Ищите alpha в диапазоне (0.1, 1.1) с шагом 0.1. 

Осуществите подбор параметра alpha по тренировочным данным (Xtrain, ytrain).

In [14]:
from sklearn.linear_model import Lasso

param_grid = {'alpha': [n * 1/10 for n in range(1,12)]}
gridsearch = GridSearchCV(Lasso(), param_grid, scoring='r2', cv=5)

gridsearch.fit(X_train, y_train)
gridsearch.cv_results_

{'mean_fit_time': array([0.00219364, 0.00059867, 0.00079865, 0.00099483, 0.0007978 ,
        0.00039897, 0.00059819, 0.00079789, 0.00039892, 0.00039902,
        0.0005981 ]),
 'std_fit_time': array([2.91712597e-03, 4.88811513e-04, 7.47679515e-04, 4.84077793e-06,
        3.98898506e-04, 4.88636178e-04, 4.88421929e-04, 3.98945848e-04,
        4.88577633e-04, 4.88694433e-04, 4.88344194e-04]),
 'mean_score_time': array([0.        , 0.00039897, 0.00039897, 0.00039892, 0.00019941,
        0.00039897, 0.00039907, 0.00039883, 0.00019946, 0.00039883,
        0.00019979]),
 'std_score_time': array([0.        , 0.00048864, 0.00048864, 0.00048858, 0.00039883,
        0.00048864, 0.00048875, 0.00048846, 0.00039892, 0.00048846,
        0.00039959]),
 'param_alpha': masked_array(data=[0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0, 1.1],
              mask=[False, False, False, False, False, False, False, False,
                    False, False, False],
        fill_value='?',
             dtype=ob

### Шаг 7.  (**0.5 балла**)
Выведите наилучший алгоритм и наилучшее качество по результатам подбора alpha (best_estimator_ и best_score_).

In [15]:
gridsearch.best_estimator_

Lasso(alpha=0.1)

In [16]:
gridsearch.best_score_

0.22915583720108373

### Шаг 8.  (**0.5 балла**)

С помощью найденного best_estimator_ сделайте предсказание на тестовых данных и выведите на экран r2-score на тесте.

In [17]:
best_pred = gridsearch.best_estimator_.predict(X_test)
r2_score(y_test, best_pred)

0.24643140462695579

### Шаг 9.  (**0.5 балла**)

Попробуем улучшить качество модели за счет добавления полиномиальных признаков. Создайте pipeline, состоящий из добавления полиномиальных признаков степени 2, а затем применения линейной регрессии.

Затем вычислите r2-score этой модели на кросс валидации с пятью фолдами.

In [18]:
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import PolynomialFeatures 
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LinearRegression

pipe = Pipeline([
    ('poly_transformer_', PolynomialFeatures(2)),
    ('scaler_', StandardScaler()),
    ('model_', LinearRegression())
    ])

R2 на каждом из фолдов:

In [19]:
cross_val_score(pipe, X, y, cv=5, scoring='r2')

array([-0.0399065 ,  0.29930017,  0.38080624,  0.30145898,  0.20882197])

Странное качество получается на первом фолде...Мб туда попали нетипичные наблюдения (?), мб зависит от параметра рандома. Из-за этого портится среднее качество тоже.

Среднее качество по всем фолдам:

In [20]:
np.mean(cross_val_score(pipe, X, y, cv=5, scoring='r2'))

0.23009617099812116

### Шаг 10.  (**0.5 балла**)
Обучите модель (pipeline) на тренировочных данных и сделайте предсказания для train и test, затем выведите на экран r2-score и MSE на тренировочных и на тестовых данных.

In [21]:
pipe.fit(X_train, y_train)

pred_train = pipe.predict(X_train)
pred_test = pipe.predict(X_test)

print('R-squared на трейне =', r2_score(y_train, pred_train),'\nR-squared на тесте = ', r2_score(y_test, pred_test))
print('MSE на трейне =', mean_squared_error(y_train, pred_train),'\nMSE на тесте = ', mean_squared_error(y_test, pred_test))

R-squared на трейне = 0.46015279277541377 
R-squared на тесте =  0.3241842009567366
MSE на трейне = 0.35559974788199566 
MSE на тесте =  0.4292808941223118


### Сделайте выводы. Для этого ответьте на вопросы: (**1 балл**)

1) Хорошее ли качество показала исходная модель (линейная регрессия без регуляризации)? Является ли эта модель переобученной?

2) Помогла ли L1-регуляризация улучшить качество модели?

3) Помогло ли добавление полиномов второй степени улучшить качество модели? Как добавление новых признаков повлияло на переобучение?

1)
Для исходной модели: MSE на трейне = 0.422, MSE на тесте =  0.409, R-squared на трейне = 0.360, R-squared на тесте =  0.357. 
Как мы знаем, сам по себе показатель MSE не имеет верхней границы и зависит от разброса переменной, поэтому судить только по нему о качестве модели нецелесообразно. Зная что зависимая переменная это какое-то целое число в диапазоне от 3 до 8, полагаю, что значение MSE=0,41 достаточно высокое для данного датасета: качество модели не очень. Об этом же говорит достаточно низкий коэффициент детерминации, смысл которого - показывать долю вариации, объясненную регрессией в общей вариации Y. То есть объясняющая способность модели достаточно низкая. 
Модель не является переобученной, скорее видим недообучение - низкие значения метрик качества на тренировочной выборке. Значения MSE и R2 на трейне не сильно превышают значения соответствующих метрик на тесте - значит не наблюдается переобучения.
На кроссвалидации среднее значение R2 еще ниже: 0.29. Это не меняет выводы.

In [22]:
sorted(pd.unique(df['quality']))

[3, 4, 5, 6, 7, 8]

2)
Регуляризация не помогла. По крайней мере для параметров alpha из указанного диапазона. Коэффициент детерминации на тесте R2=0.246, при alpha=0.1. Качество модели снизилось.

3) 
R-squared на трейне = 0.46015279277541377 

R-squared на тесте =  0.3241842009567366

MSE на трейне = 0.35559974788199566 

MSE на тесте =  0.4292808941223118

Судя по данным показателям - качество модели улучшилось! Но зато она стала более переобученной - об этом говорит увеличение разрыва между значениями метрик на трейне и тесте. R2 вообще вполне мог увеличиться только по причине увеличения количества регрессоров (тоже признак того, что модель подгоняется под данные). Нужно чекать R2-adjusted или сравнивать по информационным критериям. 

### *Попытайтесь улучшить модель (добейтесь наилучшего качества) - можно использовать любые методы.

При улучшении качества r2 на 0.1-0.2 +1 балл, при большем улучшении +2 балла (дополнительно к 5 баллам за основную часть).

Это исходная модель (с полиномиальными признаками):

In [23]:
pipe = Pipeline([
    ('poly_transformer_', PolynomialFeatures(2)),
    ('scaler_', StandardScaler()),
    ('model_', LinearRegression())
    ])
pipe.fit(X_train, y_train)
pred_test = pipe.predict(X_test)

print('R-squared на тeсте =', r2_score(y_test, pred_test))

R-squared на тeсте = 0.3241842009567366


Качество немного увеличивается, если подобрать параметр регуляризации alpha:

In [24]:
pipe = Pipeline([
    ('poly_transformer_', PolynomialFeatures(2)),
    ('scaler_', StandardScaler()),
    ('model_', Lasso(alpha=0.008, max_iter=1000))
    ])
pipe.fit(X_train, y_train)
pred_test = pipe.predict(X_test)

print('R-squared на тeсте =', r2_score(y_test, pred_test))

R-squared на тeсте = 0.3641180391633353


Просто запустила градиентный бустинг и радуюсь жизни:

In [25]:
from sklearn.ensemble import GradientBoostingRegressor
pipe = Pipeline([
    ('poly_transformer_', PolynomialFeatures(2)),
    ('scaler_', StandardScaler()),
    ('model_', GradientBoostingRegressor(learning_rate=0.1, n_estimators=500, random_state=690))
    ])
pipe.fit(X_train, y_train)
pred_test = pipe.predict(X_test)

print('R-squared на тeсте =', r2_score(y_test, pred_test))

R-squared на тeсте = 0.42658472708721273


## Часть 2. Target encoding (**всего 5 баллов**)

В этом части домашнего задания вы будете работать с выборкой `1C`. Вам нужно посчитать счетчики для `item_id` четырьмя способами:

    1) При помощи KFold схемы;  
    2) При помощи Leave-one-out схемы;
    3) При помощи smoothing схемы;
    4) При помощи expanding mean схемы.

### Подготовка данных

In [26]:
sales = pd.read_csv('sales_train.csv.gz')
sales.columns = ['date', 'date_block_num', 'shop_id', 'item_id', 'item_price', 'target']
sales

date  date_block_num  shop_id  item_id  item_price  target
0        02.01.2013               0       59    22154      999.00     1.0
1        03.01.2013               0       25     2552      899.00     1.0
2        05.01.2013               0       25     2552      899.00    -1.0
3        06.01.2013               0       25     2554     1709.05     1.0
4        15.01.2013               0       25     2555     1099.00     1.0
...             ...             ...      ...      ...         ...     ...
2935844  10.10.2015              33       25     7409      299.00     1.0
2935845  09.10.2015              33       25     7460      299.00     1.0
2935846  14.10.2015              33       25     7459      349.00     1.0
2935847  22.10.2015              33       25     7440      299.00     1.0
2935848  03.10.2015              33       25     7460      299.00     1.0

[2935849 rows x 6 columns]

In [27]:
index_cols = ['shop_id', 'item_id', 'date_block_num']

# For every month we create a grid from all shops/items combinations from that month
grid = [] 
for block_num in sales['date_block_num'].unique():
    cur_shops = sales[sales['date_block_num']==block_num]['shop_id'].unique()
    cur_items = sales[sales['date_block_num']==block_num]['item_id'].unique()
    grid.append(np.array(list(product(*[cur_shops, cur_items, [block_num]])),dtype='int32'))

#turn the grid into pandas dataframe
grid = pd.DataFrame(np.vstack(grid), columns = index_cols,dtype=np.int32)

#get aggregated values for (shop_id, item_id, month)
gb = sales.groupby(index_cols,as_index=False).agg({'target':'sum'})

#join aggregated data to the grid
all_data = pd.merge(grid,gb,how='left',on=index_cols).fillna(0)
#sort the data
all_data.sort_values(['date_block_num','shop_id','item_id'],inplace=True)

### Mean encodings без регуляризации

После проделанной технической работы, мы готовы посчитать счетчики для переменной `item_id`. 

Ниже приведены две реализации подсчета счетчиков без регуляризации. Можно использовать данный код в качестве стартовой точки для реализации различных техник регуляризации.

#### Способ 1

In [28]:
# Calculate a mapping: {item_id: target_mean}
item_id_target_mean = all_data.groupby('item_id').target.mean()

# In our non-regularized case we just *map* the computed means to the `item_id`'s
all_data['item_target_enc'] = all_data['item_id'].map(item_id_target_mean)

# Fill NaNs
all_data['item_target_enc'].fillna(0.3343, inplace=True) 

# Print correlation
encoded_feature = all_data['item_target_enc'].values
print(np.corrcoef(all_data['target'].values, encoded_feature)[0][1])

0.4830386988621791


#### Способ 2

In [29]:
'''
     Differently to `.target.mean()` function `transform` 
   will return a dataframe with an index like in `all_data`.
   Basically this single line of code is equivalent to the first two lines from of Method 1.
'''
all_data['item_target_enc'] = all_data.groupby('item_id')['target'].transform('mean')

# Fill NaNs
all_data['item_target_enc'].fillna(0.3343, inplace=True) 

# Print correlation
encoded_feature = all_data['item_target_enc'].values
print(np.corrcoef(all_data['target'].values, encoded_feature)[0][1])

0.4830386988621791


###  KFold схема (**1.25 балла**)

Необходимо реализовать Kfold схему с пятью фолдами. Используйте KFold(5) из sklearn.model_selection. 

1. Разбейте данные на 5 фолдов при помощи `sklearn.model_selection.KFold` с параметром `shuffle=False`.
2. Проитерируйтесь по фолдам: используйте 4 обучающих фолда для подсчета средних значений таргета по `item_id` и заполните этими значениями валидационный фолд на каждой итерации.

Обратите внимание на **Способ 1** из примера. В частности, изучите, как работают функции map и pd.Series.map. Они довольно полезны во многих ситуациях. 

In [30]:
from sklearn.model_selection import KFold
kf = KFold(n_splits=5, shuffle=False)

all_data['item_target_enc'] = np.nan
for train_index, val_index in kf.split(all_data):
    df_train, df_val = all_data.iloc[train_index], all_data.iloc[val_index]
    all_data.loc[all_data.index[val_index], 'item_target_enc'] = df_val['item_id'].map(df_train.groupby('item_id').target.mean())
    

all_data['item_target_enc'].fillna(0.3343, inplace=True) 
encoded_feature = all_data['item_target_enc'].values

# You will need to compute correlation like that
corr = np.corrcoef(all_data['target'].values, encoded_feature)[0][1]
print(corr)

0.41645907127988024


Ожидаемый ответ 0.4165

### Leave-one-out схема (**1.25 балла**)

Необходимо реализовать leave-one-out схему . Учтите, если вы запустите код из первого задания, задав количество фолдов такое же как размер выборки, то вы, вероятно, получите правильный ответ, но ждать будете очень-очень долго.

Для более быстрой реализации подсчета среднего таргета на всех объектах, кроме одного, вы можете:

1. Вычислить суммарный таргет по всем объектам.
2. Вычесть таргет конкретного объекта и разделить результат на `n_objects - 1`. 

Заметим, что пункт `1.` следует сделать для всех объектов. Также заметим, что пункт `2.` может быть реализован без циклов `for`.

Здесь может оказаться полезной функция .transform из **Способа 2** из примера.

In [31]:
all_data['mean_target'] = all_data.groupby('item_id')['target'].transform('mean')
sum(all_data['mean_target'])

3648206.000003528

Закомментила код по причине: я не уверена, что делаю это правильно + чтобы на этой ячейке не останавливался прогон кода (оно тут трансформится очень долго, настолько, что я не дождалась и не смогла чекнуть результат). Можете просто проигнорить

In [32]:
# all_data['mean_target'] = all_data.groupby('item_id')['target'].transform('mean')
# sum(all_data['mean_target'])

# all_data['item_target_enc'] = all_data.groupby('item_id')['target'].transform(lambda x: (sum(all_data['mean_target'])-x.mean())/(all_data.shape[0]-1))

# Print correlation
# encoded_feature = all_data['item_target_enc'].values
# corr = np.corrcoef(all_data['target'].values, encoded_feature)[0][1]
# print(corr)

??? ну, я пыталась

In [33]:
all_data['mean_target'] = all_data.groupby('item_id')['target'].transform('mean')
all_data['sum_mean_target'] = sum(all_data['mean_target']) * np.ones(all_data.shape[0])

denominator = all_data.shape[0]-1
all_data['item_target_enc'] = (all_data['sum_mean_target'] - all_data['mean_target'])/denominator

# Print correlation
encoded_feature = all_data['item_target_enc'].values
corr = np.corrcoef(all_data['target'].values, encoded_feature)[0][1]
print(corr)

-0.4830386988621776


Ожидаемый ответ 0.4803

### Smoothing (**1.25 балла**)

Необходимо реализовать smoothing с $\alpha = 100$. Используйте формулу:

$\frac{mean(target) \cdot nrows + globalmean \cdot \alpha }{nrows + \alpha}$,

где $globalmean=0.3343$. Заметим, что `nrows` - это количество объектов, принадлежащих конктертной категории, а не количество строк в датасете.

In [34]:
all_data['mean_target'] = all_data.groupby('item_id')['target'].transform('mean')
all_data['nrows'] = all_data.groupby('item_id')['target'].transform('size')

all_data['item_target_enc'] = (all_data['nrows'] * all_data['mean_target'] + 0.3343 * 100)/(all_data['nrows']+100)

encoded_feature = all_data['item_target_enc'].values
corr = np.corrcoef(all_data['target'].values, encoded_feature)[0][1]
print(corr)

0.4818198797097282


Ожидаемый ответ 0.4818

### Expanding mean схема (**1.25 балла**)

Необходимо реализовать *expanding mean* схему. Ее суть заключается в том, чтобы пройти по отсортированному в определенном порядке датасету (датасет сортируется в самом начале задания) и для подсчета счетчика для строки $m$ использовать строки от $0$ до $m-1$. Вам будет необходимо воспользоваться pandas функциями [`cumsum`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.core.groupby.DataFrameGroupBy.cumsum.html) и [`cumcount`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.core.groupby.GroupBy.cumcount.html).

In [35]:
all_data['cumsum'] = all_data.groupby('item_id')['target'].transform('cumsum')
all_data['cumcount'] = all_data.groupby('item_id')['target'].transform('cumcount') 

all_data['item_target_enc'] = (all_data['cumsum']-all_data['target']) / all_data['cumcount'] 
encoded_feature = all_data['item_target_enc'].values
all_data['item_target_enc'].fillna(0.3343, inplace=True) 

corr = np.corrcoef(all_data['target'].values, encoded_feature)[0][1]
print(corr)

0.5025245211081701


Ожидаемый ответ 0.5025